lib/srv/discovery/access_graph_azure.go

```diff
@@ -0,0 +1,409 @@
+)
+
+// reconcileAccessGraphAzure fetches Azure resources, creates a set of resources to delete and upsert based on
+// the previous fetch, and then sends the delete and upsert results to the Access Graph stream
+func (s *Server) reconcileAccessGraphAzure(
+	ctx context.Context,
+	currentTAGResources *azuresync.Resources,
+	stream accessgraphv1alpha.AccessGraphService_AzureEventsStreamClient,
+	features azuresync.Features,
+) error {
+	type fetcherResult struct {
+		result *azuresync.Resources
+		err    error
+	}
+
+	// Get all the fetchers
+	allFetchers := s.getAllTAGSyncAzureFetchers()
+	if len(allFetchers) == 0 {
+		// If there are no fetchers, we don't need to continue.
+		// We will send a delete request for all resources and return.
+		upsert, toDel := azuresync.ReconcileResults(currentTAGResources, &azuresync.Resources{})
+
+		if err := azurePush(stream, upsert, toDel); err != nil {
+			s.Log.ErrorContext(ctx, "Error pushing empty resources to TAGs", "error", err)
+		}
+		return trace.Wrap(errNoAccessGraphFetchers)
+	}
+
+	// Fetch results concurrently
+	resultsC := make(chan fetcherResult, len(allFetchers))
+	tokens := make(chan struct{}, 3)
+	accountIds := map[string]struct{}{}
+	for _, fetcher := range allFetchers {
+		fetcher := fetcher
+		accountIds[fetcher.GetSubscriptionID()] = struct{}{}
+		tokens <- struct{}{}
+		go func() {
+			defer func() {
+				<-tokens
+			}()
+			result, err := fetcher.Poll(ctx, features)
+			resultsC <- fetcherResult{result, trace.Wrap(err)}
+		}()
+	}
+
+	// Collect the results from all fetchers.
+	results := make([]*azuresync.Resources, 0, len(allFetchers))
+	errs := make([]error, 0, len(allFetchers))
+	for i := 0; i < len(allFetchers); i++ {
+		// Each fetcher can return an error and a result.
+		fetcherResult := <-resultsC
+		if fetcherResult.err != nil {
+			errs = append(errs, fetcherResult.err)
+		}
+		if fetcherResult.result != nil {
+			results = append(results, fetcherResult.result)
+		}
+	}
+
+	// Aggregate all errors into a single error.
+	err := trace.NewAggregate(errs...)
+	if err != nil {
+		s.Log.ErrorContext(ctx, "Error polling TAGs", "error", err)
+	}
+	result := azuresync.MergeResources(results...)
+
+	// Merge all results into a single result
+	upsert, toDel := azuresync.ReconcileResults(currentTAGResources, result)
+	pushErr := azurePush(stream, upsert, toDel)
+
+	if pushErr != nil {
+		s.Log.ErrorContext(ctx, "Error pushing TAGs", "error", pushErr)
+		return nil
+	}
+
+	// Update the currentTAGResources with the result of the reconciliation.
+	*currentTAGResources = *result
+	return nil
+}
+
+// azurePushUpsertInBatches upserts resources to the Access Graph in batches
+func azurePushUpsertInBatches(
+	client accessgraphv1alpha.AccessGraphService_AzureEventsStreamClient,
+	upsert *accessgraphv1alpha.AzureResourceList,
+) error {
+	for i := 0; i < len(upsert.Resources); i += batchSize {
+		end := i + batchSize
+		if end > len(upsert.Resources) {
+			end = len(upsert.Resources)
+		}
```

Comment:
I think we can use this instead.
https://pkg.go.dev/slices@master#Chunk

Reply:
I don't think we use go 1.23 everywhere